File: CS/BookmarkInsertLR/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraRichEdit.API.Native;

namespace BookmarkInsertLR {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e) {
            richEditControl1.CreateNewDocument();
            richEditControl1.Options.Bookmarks.Visibility = DevExpress.XtraRichEdit.RichEditBookmarkVisibility.Visible;
            Document document = richEditControl1.Document;
            document.InsertText(document.CreatePosition(0), "Bookmark 1Bookmark 2Bookmark 3Bookmark 4");
            document.CreateBookmark(document.CreateRange(0, 10), "b1");
            document.CreateBookmark(document.CreateRange(10, 10), "b2");
            document.CreateBookmark(document.CreateRange(20, 10), "b3");
            document.CreateBookmark(document.CreateRange(30, 10), "b4");
            document.CreateBookmark(document.CreateRange(5, 15), "b5");
            cbBookmarks.Items.AddRange(new string[] { "b1", "b2", "b3", "b4", "b5" });

        }
        int count = 0;
        private void btnLeft_Click(object sender, EventArgs e) {
            count++;
            Document document = richEditControl1.Document;
            document.BeginUpdate();
            DocumentRange textRange = document.InsertText(document.CaretPosition, "L" + count.ToString());
            CorrectBookmarksStart(textRange.Start, textRange.End);
            CorrectBookmarksEnd(textRange.Start, textRange.End);
            document.EndUpdate();
        }

        private void btnRight_Click(object sender, EventArgs e) {
            count++;
            Document document = richEditControl1.Document;
            document.BeginUpdate();
            DocumentRange textRange = document.InsertText(document.CaretPosition, "R" + count.ToString());
            CorrectBookmarksStart(textRange.End, textRange.Start);
            CorrectBookmarksEnd(textRange.End, textRange.Start);
            document.EndUpdate();
            
        }

        private void btnBetween_Click(object sender, EventArgs e) {
            count++;
            Document document = richEditControl1.Document;
            document.BeginUpdate();
            DocumentRange textRange = document.InsertText(document.CaretPosition, "B" + count.ToString());
            CorrectBookmarksStart(textRange.Start, textRange.End);
            CorrectBookmarksEnd(textRange.End, textRange.Start);
            document.EndUpdate();
            
        }
        void CorrectBookmarksStart(DocumentPosition oldBookmarkStart, DocumentPosition newBookmarkStart) {
            Document document = richEditControl1.Document;
            document.BeginUpdate();
            List<Bookmark> bookmarksAtRight = new List<Bookmark>();
            foreach (Bookmark bookmark in document.Bookmarks) {
                if (bookmark.Range.Start.ToInt() == oldBookmarkStart.ToInt())
                    bookmarksAtRight.Add(bookmark);
            }
            foreach (Bookmark bookmark in bookmarksAtRight) {
                int bookmarkEnd = bookmark.Range.End.ToInt();
                string bookmarkName = bookmark.Name;
                document.RemoveBookmark(bookmark);
                document.CreateBookmark(newBookmarkStart, bookmarkEnd - newBookmarkStart.ToInt(), bookmarkName);
            }
            document.EndUpdate();
        }
        void CorrectBookmarksEnd(DocumentPosition oldBookmarkEnd, DocumentPosition newBookmarkEnd) {
            Document document = richEditControl1.Document;
            document.BeginUpdate();
            List<Bookmark> bookmarksAtLeft = new List<Bookmark>();
            foreach (Bookmark bookmark in document.Bookmarks) {
                if (bookmark.Range.End.ToInt() == oldBookmarkEnd.ToInt())
                    bookmarksAtLeft.Add(bookmark);
            }
            foreach (Bookmark bookmark in bookmarksAtLeft) {
                int bookmarkStart = bookmark.Range.Start.ToInt();
                string bookmarkName = bookmark.Name;
                document.RemoveBookmark(bookmark);
                document.CreateBookmark(document.CreatePosition(bookmarkStart), newBookmarkEnd.ToInt() - bookmarkStart, bookmarkName);
            }
            document.EndUpdate();
        }

        private void cbBookmarks_SelectedValueChanged(object sender, EventArgs e) {
            Document document = richEditControl1.Document;
            string value = cbBookmarks.SelectedItem as String;
            if(String.IsNullOrEmpty(value))
                return;
            document.SelectBookmark(document.Bookmarks[value]);
        }
    }
}

File: VB/CS/BookmarkInsertLR/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraRichEdit.API.Native

Namespace BookmarkInsertLR

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            richEditControl1.CreateNewDocument()
            richEditControl1.Options.Bookmarks.Visibility = DevExpress.XtraRichEdit.RichEditBookmarkVisibility.Visible
            Dim document As Document = richEditControl1.Document
            document.InsertText(document.CreatePosition(0), "Bookmark 1Bookmark 2Bookmark 3Bookmark 4")
            document.Bookmarks.Create(document.CreateRange(0, 10), "b1")
            document.Bookmarks.Create(document.CreateRange(10, 10), "b2")
            document.Bookmarks.Create(document.CreateRange(20, 10), "b3")
            document.Bookmarks.Create(document.CreateRange(30, 10), "b4")
            document.Bookmarks.Create(document.CreateRange(5, 15), "b5")
            cbBookmarks.Items.AddRange(New String() {"b1", "b2", "b3", "b4", "b5"})
        End Sub

        Private count As Integer = 0

        Private Sub btnLeft_Click(ByVal sender As Object, ByVal e As EventArgs)
            count += 1
            Dim document As Document = richEditControl1.Document
            document.BeginUpdate()
            Dim textRange As DocumentRange = document.InsertText(document.CaretPosition, "L" & count.ToString())
            CorrectBookmarksStart(textRange.Start, textRange.End)
            CorrectBookmarksEnd(textRange.Start, textRange.End)
            document.EndUpdate()
        End Sub

        Private Sub btnRight_Click(ByVal sender As Object, ByVal e As EventArgs)
            count += 1
            Dim document As Document = richEditControl1.Document
            document.BeginUpdate()
            Dim textRange As DocumentRange = document.InsertText(document.CaretPosition, "R" & count.ToString())
            CorrectBookmarksStart(textRange.End, textRange.Start)
            CorrectBookmarksEnd(textRange.End, textRange.Start)
            document.EndUpdate()
        End Sub

        Private Sub btnBetween_Click(ByVal sender As Object, ByVal e As EventArgs)
            count += 1
            Dim document As Document = richEditControl1.Document
            document.BeginUpdate()
            Dim textRange As DocumentRange = document.InsertText(document.CaretPosition, "B" & count.ToString())
            CorrectBookmarksStart(textRange.Start, textRange.End)
            CorrectBookmarksEnd(textRange.End, textRange.Start)
            document.EndUpdate()
        End Sub

        Private Sub CorrectBookmarksStart(ByVal oldBookmarkStart As DocumentPosition, ByVal newBookmarkStart As DocumentPosition)
            Dim document As Document = richEditControl1.Document
            document.BeginUpdate()
            Dim bookmarksAtRight As List(Of Bookmark) = New List(Of Bookmark)()
            For Each bookmark As Bookmark In document.Bookmarks
                If bookmark.Range.Start.ToInt() = oldBookmarkStart.ToInt() Then bookmarksAtRight.Add(bookmark)
            Next

            For Each bookmark As Bookmark In bookmarksAtRight
                Dim bookmarkEnd As Integer = bookmark.Range.End.ToInt()
                Dim bookmarkName As String = bookmark.Name
                document.Bookmarks.Remove(bookmark)
                document.Bookmarks.Create(newBookmarkStart, bookmarkEnd - newBookmarkStart.ToInt(), bookmarkName)
            Next

            document.EndUpdate()
        End Sub

        Private Sub CorrectBookmarksEnd(ByVal oldBookmarkEnd As DocumentPosition, ByVal newBookmarkEnd As DocumentPosition)
            Dim document As Document = richEditControl1.Document
            document.BeginUpdate()
            Dim bookmarksAtLeft As List(Of Bookmark) = New List(Of Bookmark)()
            For Each bookmark As Bookmark In document.Bookmarks
                If bookmark.Range.End.ToInt() = oldBookmarkEnd.ToInt() Then bookmarksAtLeft.Add(bookmark)
            Next

            For Each bookmark As Bookmark In bookmarksAtLeft
                Dim bookmarkStart As Integer = bookmark.Range.Start.ToInt()
                Dim bookmarkName As String = bookmark.Name
                document.Bookmarks.Remove(bookmark)
                document.Bookmarks.Create(document.CreatePosition(bookmarkStart), newBookmarkEnd.ToInt() - bookmarkStart, bookmarkName)
            Next

            document.EndUpdate()
        End Sub

        Private Sub cbBookmarks_SelectedValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim document As Document = richEditControl1.Document
            Dim value As String = TryCast(cbBookmarks.SelectedItem, String)
            If String.IsNullOrEmpty(value) Then Return
            document.Bookmarks.Select(document.Bookmarks(value))
        End Sub
    End Class
End Namespace